### PR TITLE
[AOT] fix: crash when setting uniform of particle system

### PR DESCRIPTION
**Pull request by PPpro**
Re: https://github.com/cocos/3d-tasks/issues/16803

### Changelog

* fix crash when setting uniform of particle system

-------

### Continuous Integration

This pull request:

* [ ] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

**Comment by minggo** [comment]
Please add more detail information:
- what caused crash?
- how to fix it?

**Comment by PPpro** [comment]
@minggo detailed commet added

File: cocos/particle/renderer/particle-system-renderer-cpu.ts [hunk 1]
```diff
@@ -339,17 +340,20 @@ export default class ParticleSystemRendererCPU extends ParticleSystemRendererBas
...
-            this._particleSystem?.node.getWorldScale(this._node_scale);
+            this._particleSystem?.node.getWorldScale(nodeScale);
             break;
         default:
             break;
         }
-        pass.setUniform(this._uScaleHandle, this._node_scale);
+        // NOTE: the `_node_scale` should be a Vec3, but we implement `scale` uniform property as a Vec4,
+        // here we pass a temperate Vec4 object to prevent creating Vec4 object every time we set uniform.
+        pass.setUniform(this._uScaleHandle, _tempNodeScale.set(nodeScale.x, nodeScale.y, nodeScale.z));
```
Comment (minggo) [inline cocos/particle/renderer/particle-system-renderer-cpu.ts:356]: Why not change to use vec3 in uniform too?

File: cocos/particle/renderer/particle-system-renderer-gpu.ts [hunk 2]
```diff
@@ -295,17 +296,20 @@ export default class ParticleSystemRendererGPU extends ParticleSystemRendererBas
...
-            this._particleSystem.node.getWorldScale(this._node_scale);
+            this._particleSystem.node.getWorldScale(nodeScale);
             break;
         default:
             break;
         }
-        pass.setUniform(pass.getHandle('scale'), this._node_scale);
+        // NOTE: the `_node_scale` should be a Vec3, but we implement `scale` uniform property as a Vec4,
+        // here we pass a temperate Vec4 object to prevent creating Vec4 object every time we set uniform.
+        pass.setUniform(pass.getHandle('scale'), _tempNodeScale.set(nodeScale.x, nodeScale.y, nodeScale.z));
```
Comment (minggo) [inline cocos/particle/renderer/particle-system-renderer-gpu.ts:312]: The same as above.

File: cocos/particle/renderer/particle-system-renderer-cpu.ts (hunk 1 above)
Comment (PPpro) [reply cocos/particle/renderer/particle-system-renderer-cpu.ts:356]: I've checked the shader implementation, it relies on Vec4's w property

File: cocos/particle/renderer/particle-system-renderer-gpu.ts (hunk 2 above)
Comment (PPpro) [reply cocos/particle/renderer/particle-system-renderer-gpu.ts:312]: the same

File: cocos/particle/renderer/particle-system-renderer-cpu.ts (hunk 1 above)
Comment (minggo) [reply cocos/particle/renderer/particle-system-renderer-cpu.ts:356]: Shader will automatically add w value.

Comment (PPpro) [reply cocos/particle/renderer/particle-system-renderer-cpu.ts:356]: but then engine crash on the validator

**Review by minggo — APPROVED** [review]

**Review by zxx43 — APPROVED** [review]

**Review by SantyWang — APPROVED** [review]

**Pull request merged**